junit-jupiter-api/src/main/java/org/junit/jupiter/api/extension/StoredValues.java

```diff
@@ -0,0 +1,153 @@
+
+package org.junit.jupiter.api.extension;
+
+import static org.junit.platform.commons.meta.API.Usage.Experimental;
+
+import java.util.Objects;
+import java.util.function.Function;
+
+import org.junit.jupiter.api.extension.ExtensionContext.Store;
+import org.junit.platform.commons.meta.API;
+
+/**
+ * Provides isolated, type-safe access to values in a {@link Store}.
+ *
+ * <p>Example use:
+ *
+ * <pre>{@code
+ * public class TimingExtension
+ *     implements BeforeTestExecutionCallback, AfterTestExecutionCallback {
+ *   private static final StoredValues<Method, Instant> START_TIMES
+ *       = new StoredValues<>();
+ * 
+ *   @Override
+ *   public void beforeTestExecution(ExtensionContext context) {
+ *     START_TIMES.put(
+ *         getStore(context), context.getRequiredTestMethod(), Instant.now());
+ *   }
+ * }
+ * }</pre>
```

Comment:
I feel the same as @marcphilipp: I don't (yet) see how this is an improvement.

It appears to be just a _different_ set of hoops to jump through.

Plus, with this proposal, the static `StoredValues` is limited to storing only instances of `Instant`. Thus if the extension needed to have additional types of data stored between callbacks, the extension would then have to have multiple instances of `StoredValues`, one for each concrete type of thing to store. So that honestly seems like a step back to me.

Or am I missing something?